### PR TITLE
install_tree relative symlinks

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes https://github.com/spack/spack/issues/9001

See: https://github.com/spack/spack/pull/8289

This updates the `install_tree` method and its tests:

* `copy_tree`/`install_tree` now preserve relative link targets (if the symlink in the source directory structure is relative, the symlink created in the destination will be relative)
* Added support for `ignore` argument back to `copy_tree`/`install_tree`. It is no longer the object output by `shutil.ignore_patterns`: you pass a function that accepts a path relative to the source and returns whether that path should be copied.
* When a symlink target is absolute, `copy_tree` and `install_tree` now rewrite the source prefix to be the destination prefix 
* `copy_tree` tests no longer check permissions: `copy_tree` doesn't enforce anything about permissions so its tests don't check for that
* `install_tree` tests no longer check for exact permission matching since it can add file permissions

### CODE REVIEW

File: lib/spack/llnl/util/filesystem.py [hunk 1]
```diff
@@ -323,30 +337,43 @@ def copy_tree(src, dest, symlinks=True, _permissions=False):
         symlinks (bool): whether or not to preserve symlinks
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: Need to add `ignore`

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -323,30 +337,43 @@ def copy_tree(src, dest, symlinks=True, _permissions=False):
...
         _permissions (bool): for internal use only
     """
-    if _permissions:
-        tty.debug('Installing {0} to {1}'.format(src, dest))
-    else:
-        tty.debug('Copying {0} to {1}'.format(src, dest))
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: Why remove the debug messages?

File: lib/spack/llnl/util/filesystem.py [hunk 3]
```diff
@@ -357,7 +384,7 @@ def install_tree(src, dest, symlinks=True):
         dest (str): the destination directory
         symlinks (bool): whether or not to preserve symlinks
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: Need to add `ignore`

File: lib/spack/llnl/util/filesystem.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/llnl/util/filesystem.py]: done

File: lib/spack/llnl/util/filesystem.py (hunk 3 above)
Comment (scheibelp) [reply lib/spack/llnl/util/filesystem.py]: done

**Comment by tgamblin** [comment]
Thanks @scheibelp!

**Pull request merged**